src/audio/chain_dma.c

```diff
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
 	rfree(dev);
 }
 
+static enum task_state chain_task_run(void *data)
+{
+	struct dma_status stat;
+	size_t link_avail_bytes, link_free_bytes, host_avail_bytes, host_free_bytes;
+	uint32_t link_read_pos, host_read_pos;
+	struct chain_dma_data *cd = (struct chain_dma_data *)data;
+
+	/* retrieve current status of link DMA */
+	int ret = dma_get_status(cd->chan_link->dma->z_dev, cd->chan_link->index,
+				 &stat);
+	if (ret < 0)
+		tr_err(&chain_dma_tr, "chain_task_run(): dma_get_status() failed, ret = %u",
+			 ret);
+
+	link_avail_bytes = stat.pending_length;
+	link_free_bytes = stat.free;
+	link_read_pos = stat.read_position;
+
+
+	/* retrieve current status of host DMA */
+	ret = dma_get_status(cd->chan_host->dma->z_dev, cd->chan_host->index, &stat);
+	if (ret < 0)
+		tr_err(&chain_dma_tr, "chain_task_run(): dma_get_status() failed, ret = %u",
+		       ret);
+
+	host_avail_bytes = stat.pending_length;
+	host_free_bytes = stat.free;
+	host_read_pos = stat.read_position;
+
+	uint32_t link_type = cd->link_connector_node_id.f.dma_type;
+
+	if (link_type == ipc4_hda_link_input_class) {
+		const size_t increment = MIN(host_free_bytes, link_avail_bytes);
+
+		ret = dma_reload(cd->chan_host->dma->z_dev, cd->chan_host->index,
+				 0, 0, increment);
+		if (ret < 0)
+			tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() host failed, ret = %u",
+			       ret);
+
+		ret = dma_reload(cd->chan_link->dma->z_dev, cd->chan_link->index, 0, 0,
+				 increment);
+		if (ret < 0)
+			tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() link failed, ret = %u",
+			       ret);
+	} else {
+		const size_t half_buff_size = cd->dma_buffer->stream.size / 2;
+
+		if ((!cd->first_data_received) && (host_avail_bytes > half_buff_size)) {
+			ret = dma_reload(cd->chan_link->dma->z_dev,
+					 cd->chan_link->index, 0, 0,
+					 half_buff_size);
+			if (ret < 0)
+				tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() link failed, ret = %u",
+				       ret);
+			cd->first_data_received = true;
+
+		} else if (cd->first_data_received) {
+			const size_t transferred = chain_get_transferred_data_size(
+						   link_read_pos, host_read_pos,
+						   cd->dma_buffer->stream.size);
+
+			ret = dma_reload(cd->chan_host->dma->z_dev, cd->chan_host->index,
+					 0, 0, transferred);
+			if (ret < 0)
+				tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() host failed, ret = %u",
+				       ret);
+
+			if ((host_avail_bytes >= half_buff_size) &&
+			    (link_free_bytes >= half_buff_size)) {
+				ret = dma_reload(cd->chan_link->dma->z_dev, cd->chan_link->index,
+						 0, 0, half_buff_size);
+				if (ret < 0)
+					tr_err(&chain_dma_tr, "chain_task_run(): dma_reload() link failed, ret = %u",
+					       ret);
+			}
+		}
+	}
+
+	return SOF_TASK_STATE_RESCHEDULE;
+}
+
+int chain_link_start(struct comp_dev *dev)
+{
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	int err = dma_start(cd->chan_link->dma->z_dev, cd->chan_link->index);
+
+	if (err < 0)
+		return err;
```

Comment:
this is hard to read like this, separate the declaration and the call to dma_start().